_posts/2022-9-5-Diplomat @ Google.md

```diff
@@ -0,0 +1,503 @@
+conversion cost.
+4. Boxed opaques can only be returned, but not accepted as a parameters because
+most other languages don't have a way to give up ownership.
+
+On the other hand, structs not marked with the `#[diplomat::opaque]` attribute
+have nearly opposite properties:
+1. Fields are public to the other language.
+2. Fields can only be primitive types like numbers, booleans, characters,
+owned opaques, and other structs. My project extended this to also allow for
+borrowed opaques, string slices, and slices of primitives.
+3. Can only cross the FFI boundary by value, meaning that every field is also
+converted at the boundary.
+
+Additionally, Diplomat also supports fieldless enums, which are just passed
+as their discriminents across the boundary and then converted appropriately.
+
+# My Intern Project
+
+Rust's type system allows for expressing mutability, ownership and lifetimes,
+concepts that most other languages simply do not have. My work this summer was
+to extend Diplomat's abstract model to figure out how we can take advantage of
+Rust's precise control over these factors across the FFI boundary, particularly
+with returning references to borrowed data.
+
+One of the problems that Rust's lifetime system solves is dangling pointers,
+and it does this statically by preventing references from outliving their owners.
+But rustc can only do static analysis on pure Rust code, and this is the main
+problem with returning pointers across FFI (aside from XOR aliasing, which is
+[being worked on](https://github.com/rust-diplomat/diplomat/issues/225)).
+
+The solution is to instead make Diplomat enforce whatever technique the other
+language uses to prevent dangling references. In garbage collected languages
+like JavaScript, this involves attaching an edge to the owner to stop it from
+being GCd early. In manually memory-managed languages like C++, this involves
+generating documentation notifying the programmer of what lifetime relationships
+must be upheld.
+
+Let's walk through a simple example:
+```rust
+#[diplomat::bridge]
+mod ffi {
+    #[diplomat::opaque]
+    pub struct Bar {
+        // ...
+    }
+
+    #[diplomat::opaque]
+    pub struct Foo<'a> {
+        bar: &'a Bar,
+    }
+
+    impl<'a> Foo<'a> {
+        pub fn get_bar(&self) -> &'a Bar {
+            &self.bar
+        }
+    }
+}
+```
+When Diplomat sees `Foo::get_bar`, it sees the expanded `self: &Foo<'a>`
+parameter and the returned `&'a Bar` type, and infers that the output shouldn't
+outlive the `&self` parameter because they both have `'a`.
+
+In the case that this API requires JavaScript bindings, Diplomat would use this
+lifetime relationship to attach an edge from the output to the input that it
+borrows from to generate an API that resembles the following:
+```js
+class Foo {
+    // ...
+
+    get_bar() {
+        // `this.underlying` is the ptr to the Rust opaque
+        const bar = wasm.Foo_get_bar(this.underlying);
+        bar.fooRef = this;
```

Comment:
nit: comment saying that we "add an edge" so that the GC doesn't collect it.

Not everyone will realize what's going on here with the GC